petab/v2/lint.py

```diff
@@ -0,0 +1,549 @@
+"""Validation of PEtab problems"""
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import IntEnum
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+
+from petab.C import NOISE_PARAMETERS, OBSERVABLE_PARAMETERS
+from petab.v1 import (
+    assert_model_parameters_in_condition_or_parameter_table,
+)
+from petab.v1.C import (
+    ESTIMATE,
+    MODEL_ENTITY_ID,
+    NOMINAL_VALUE,
+    PARAMETER_DF_REQUIRED_COLS,
+    PARAMETER_ID,
+)
+from petab.v1.conditions import get_parametric_overrides
+from petab.v1.lint import (
+    _check_df,
+    assert_no_leading_trailing_whitespace,
+    assert_parameter_bounds_are_numeric,
+    assert_parameter_estimate_is_boolean,
+    assert_parameter_id_is_string,
+    assert_parameter_prior_parameters_are_valid,
+    assert_parameter_prior_type_is_valid,
+    assert_parameter_scale_is_valid,
+    assert_unique_parameter_ids,
+    check_ids,
+    check_parameter_bounds,
+)
+from petab.v1.measurements import split_parameter_replacement_list
+from petab.v1.observables import get_output_parameters, get_placeholders
+from petab.v1.parameters import (
+    get_valid_parameters_for_parameter_table,
+)
+
+from ..v1 import (
+    assert_measurement_conditions_present_in_condition_table,
+    check_condition_df,
+    check_measurement_df,
+    check_observable_df,
+)
+from .problem import Problem
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "ValidationIssueSeverity",
+    "ValidationIssue",
+    "ValidationResultList",
+    "ValidationError",
+    "ValidationTask",
+    "CheckModel",
+    "CheckTableExists",
+    "CheckMeasurementTable",
+    "CheckConditionTable",
+    "CheckObservableTable",
+    "CheckParameterTable",
+    "CheckAllParametersPresentInParameterTable",
+    "CheckValidParameterInConditionOrParameterTable",
+    "CheckVisualizationTable",
+    "lint_problem",
+    "default_validation_tasks",
+]
+
+
+class ValidationIssueSeverity(IntEnum):
+    """The severity of a validation issue."""
+
+    # INFO: Informational message, no action required
+    INFO = 10
+    # WARNING: Warning message, potential issues
+    WARNING = 20
+    # ERROR: Error message, action required
+    ERROR = 30
+    # CRITICAL: Critical error message, stops further validation
+    CRITICAL = 40
+
+
+@dataclass
+class ValidationIssue:
+    """The result of a validation task.
+
+    Attributes:
+        level: The level of the validation event.
+        message: The message of the validation event.
+    """
+
+    level: ValidationIssueSeverity
+    message: str
+
+    def __str__(self):
+        return f"{self.level.name}: {self.message}"
+
+
+class ValidationError(ValidationIssue):
+    """A validation result with level ERROR."""
+
+    def __init__(self, message: str):
+        super().__init__(ValidationIssueSeverity.ERROR, message)
```

Comment:
But then we'd still have to override `__init__`, otherwise one can do `ValidationError(ValidationIssueSeverity.INFO, "bla")` which would be weird.

I think I'd rather stick to `__init__`, but validating `level` there would be good.

Reply:
This would remove it from `__init__`
```python
@dataclass
class ValidationError(ValidationIssue):
    level: ValidationIssueSeverity = field(default=ValidationIssueSeverity.ERROR, init=False)
```

but also fine as is